File: library/src/main/java/me/proxer/library/entity/anime/Stream.java
package me.proxer.library.entity.anime;

import com.squareup.moshi.Json;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Value;
import me.proxer.library.entity.ProxerDateItem;
import me.proxer.library.entity.ProxerIdItem;
import me.proxer.library.entity.ProxerImageItem;

import javax.annotation.Nullable;
import java.util.Date;

/**
 * Entity representing a single streaming option of an anime.
 *
 * @author Ruben Gees
 */
@Value
@EqualsAndHashCode(onParam = @__({@Nullable}))
public class Stream implements ProxerIdItem, ProxerImageItem, ProxerDateItem {

    /**
     * {@inheritDoc}
     */
    @Getter(onMethod = @__({@Override}))
    @Json(name = "id")
    private String id;

    /**
     * Returns the type of hoster of this stream.
     */
    @Json(name = "type")
    private String hoster;

    /**
     * Returns the name of the hoster.
     */
    @Json(name = "name")
    private String hosterName;

    /**
     * {@inheritDoc}
     */
    @Getter(onMethod = @__({@Override}))
    @Json(name = "img")
    private String image;

    /**
     * Returns the id of the uploader.
     */
    @Json(name = "uploader")
    private String uploaderId;

    /**
     * Returns the username of the uploader.
     */
    @Json(name = "username")
    private String uploaderName;

    /**
     * {@inheritDoc}
     */
    @Getter(onMethod = @__({@Override}))
    @Json(name = "timestamp")
    private Date date;

    /**
     * Returns the id of the translator group, if present.
     */
    @Nullable
    @Getter
    @Json(name = "tid")
    private String translatorGroupId;

    /**
     * Returns the name of the translator group, if present.
     */
    @Nullable
    @Getter
    @Json(name = "tname")
    private String translatorGroupName;

    /**
     * Returns if the stream is hosted at an official hoster like Crunchyroll.
     */
    @Getter
    @Json(name = "legal")
    public boolean isOfficial;
}

File: library/src/main/kotlin/me/proxer/library/entity/anime/Stream.kt
package me.proxer.library.entity.anime

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import me.proxer.library.entity.ProxerDateItem
import me.proxer.library.entity.ProxerIdItem
import me.proxer.library.entity.ProxerImageItem
import me.proxer.library.internal.adapter.NumberBasedBoolean
import java.util.Date

/**
 * Entity representing a single streaming option of an anime.
 *
 * @property hoster The type of hoster of this stream.
 * @property hosterName The name of the hoster.
 * @property uploaderId The id of the uploader.
 * @property uploaderName The username of the uploader.
 * @property translatorGroupId The id of the translator group, if present.
 * @property translatorGroupName The name of the translator group, if present.
 * @property isOfficial If the stream is hosted at an official hoster like Crunchyroll.
 * @property isPublic If the stream is public and should be shown to guests.
 *
 * @author Ruben Gees
 */
@JsonClass(generateAdapter = true)
data class Stream(
    @Json(name = "id") override val id: String,
    @Json(name = "type") val hoster: String,
    @Json(name = "name") val hosterName: String,
    @Json(name = "img") override val image: String,
    @Json(name = "uploader") val uploaderId: String,
    @Json(name = "username") val uploaderName: String,
    @Json(name = "timestamp") override val date: Date,
    @Json(name = "tid") val translatorGroupId: String?,
    @Json(name = "tname") val translatorGroupName: String?,
    @field:NumberBasedBoolean @Json(name = "legal") val isOfficial: Boolean,
    @field:NumberBasedBoolean @Json(name = "public") val isPublic: Boolean
) : ProxerIdItem, ProxerImageItem, ProxerDateItem
